controllers/mirroring/mirroring_controller.go

```diff
@@ -0,0 +1,478 @@
+	enqueueConfigMapRequest := handler.EnqueueRequestsFromMapFunc(
+		func(_ context.Context, obj client.Object) []reconcile.Request {
+			return []reconcile.Request{{
+				NamespacedName: types.NamespacedName{
+					Name:      util.StorageClientMappingConfigName,
+					Namespace: obj.GetNamespace(),
+				},
+			}}
+		},
+	)
+
+	generationChangePredicate := builder.WithPredicates(predicate.GenerationChangedPredicate{})
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(
+			&corev1.ConfigMap{},
+			builder.WithPredicates(util.NamePredicate(util.StorageClientMappingConfigName)),
+		).
+		Watches(
+			&ocsv1.StorageClusterPeer{},
+			enqueueConfigMapRequest,
+			generationChangePredicate,
+		).
+		Watches(
+			&ocsv1alpha1.StorageConsumer{},
+			enqueueConfigMapRequest,
+			generationChangePredicate,
+		).
+		Watches(
+			&rookCephv1.CephRBDMirror{},
+			enqueueConfigMapRequest,
+			generationChangePredicate,
+		).
+		Watches(
+			&rookCephv1.CephBlockPool{},
+			enqueueConfigMapRequest,
+			generationChangePredicate,
+		).
+		Watches(
+			&rookCephv1.CephBlockPoolRadosNamespace{},
+			enqueueConfigMapRequest,
+			generationChangePredicate,
+		).
+		Complete(r)
+}
+
+//+kubebuilder:rbac:groups=ocs.openshift.io,resources=storageclusterpeers;storageconsumers,verbs=get;list;watch
+//+kubebuilder:rbac:groups=core,resources=configmaps,verbs=get;list;watch
+//+kubebuilder:rbac:groups=core,resources=configmaps/finalizers,verbs=update
+//+kubebuilder:rbac:groups=core,resources=secrets,verbs=get;list;watch;create;update;delete
+//+kubebuilder:rbac:groups=ceph.rook.io,resources=cephrbdmirrors,verbs=get;list;watch;create;update;delete
+//+kubebuilder:rbac:groups=ceph.rook.io,resources=cephblockpools;cephblockpoolradosnamespaces,verbs=get;list;watch;create;update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+func (r *MirroringReconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.Result, error) {
+	r.ctx = ctx
+	r.log = log.FromContext(ctx, "StorageClientMapping", request.NamespacedName)
+	r.log.Info("Starting reconcile")
+
+	clientMappingConfig := &corev1.ConfigMap{}
+	clientMappingConfig.Name = request.Name
+	clientMappingConfig.Namespace = request.Namespace
+
+	if err := r.get(clientMappingConfig); err != nil {
+		if k8serrors.IsNotFound(err) {
+			r.log.Info("ConfigMap %s not found. Ignoring since object must be deleted.")
+			return ctrl.Result{}, nil
+		}
+		r.log.Error(err, "Failed to get ConfigMap.")
+		return ctrl.Result{}, err
+	}
+	return r.reconcilePhases(clientMappingConfig)
+}
+
+func (r *MirroringReconciler) reconcilePhases(clientMappingConfig *corev1.ConfigMap) (ctrl.Result, error) {
+
+	disableMirroring := false
+	disableMirroring = disableMirroring || !clientMappingConfig.DeletionTimestamp.IsZero()
+	disableMirroring = disableMirroring || clientMappingConfig.Data == nil || len(clientMappingConfig.Data) < 1
```

Comment:
1. Prefer the positive case for the negative case, it removes negations from the main code path
2. The var name should reflect the state, not the request

```suggestion
	shouldMirror := clientMappingConfig.DeletionTimestamp.IsZero() &&
		clientMappingConfig.Data == nil && 
		len(clientMappingConfig.Data) < 1
```